src/coreclr/jit/morph.cpp

```diff
@@ -8135,6 +8136,34 @@ GenTree* Compiler::fgMorphSmpOp(GenTree* tree, MorphAddrContext* mac, bool* optA
                         }
                     }
                 }
+
+                // Optimization for:
+                //    a & c ==/!= c
+                // =>
+                //    ~a & c ==/!= 0
+                // where c is a constant
+                // The optimization could also take the form of a & ~c but the comparison would have to switch between EQ to NE/NE to EQ
+                if (op1->OperIs(GT_AND) && varTypeIsIntegral(op1) && op2->IsIntegralConst())
+                {
+                    GenTree* andOp1 = op1->AsOp()->gtOp1;
+                    GenTree* andOp2 = op1->AsOp()->gtOp2;
+
+                    ssize_t cnsVal = op2->AsIntCon()->IconValue();
+
+                    if (andOp1->TypeIs(TYP_INT, TYP_LONG) && andOp2->IsIntegralConst() && andOp2->AsIntCon()->IconValue() == cnsVal)
```

Comment:
IsIntegralConst should be replaced with IsIntCns or whatever it's called, I forgot. Otherwise `AsIntCon()` might fails for what is actually `AsLngCon`

Reply:
Also, you can just do `GenTree::Compare(op2, andOp2)`